### PR TITLE
Standalone cluster testing process

### LINKED ISSUE #413: Create a process for Spark cluster testing
## Description

We should have a process on how to create a multi-node Spark cluster and test against it. This will be useful to simulate a real-world use case where Spark applications are submitted to a remote cluster. 

Note that we should take into consideration pyspark environment, in particular Jupyter Notebook since it is a popular use case. 

Reason: Improves developer debugging by aligning dev test environment with typical user environment

### LINKED ISSUE #466: Allow standalone clusters to run with different Spark configurations.
## Description

Currently, we have a standalone cluster running inside our docker environment. This is used for automated testing on GitHub, as well as providing a distributed environment similar to real-world use cases of our connector. 

However, there's currently no documentation on how to change the Spark/ Hadoop version, and we would like this process to be as automated as possible.

May possibly be part of the scope for #303

### PR DESCRIPTION
**Pull request by Aryex**
### Summary

The PR made changes to facilitate testing the connector against a distributed environment. 

### Changes

Docker environment:

- Added a spark standalone cluster using bitnami's Spark container image. This allows us to easily change the Spark install and worker count.
- Updated `docker_client_1` Dockerfile to be based on bitnami's Spark containers to make use of the Spark install.
  - Both `docker_client_1` and cluster Spark version are synced
- `sandbox-clientenv` scripts are updated with options to change the worker count.

Functional testing
- Updated functional tests with option `-r` to set up test suites for use with `spark-submit`.
- `submit-functional-tests.sh` updated to accommodate changes. Also added support to pass in all parameters.
- Workflow updated.

Other
- Nightly workflow execution schedule updated from Sunday - Thursday to Monday - Friday.

### Related Issue

Close #413 
Close #466 

### Additional Reviewers

@alexey-temnikov 
@alexr-bq 
@jeremyp-bq 
@jonathanl-bq

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
-    wget https://archive.apache.org/dist/hadoop/common/hadoop-3.3.0/hadoop-3.3.0.tar.gz && \
-    tar xvf spark-3.1.2-bin-without-hadoop.tgz && \
-    tar xvf hadoop-3.3.0.tar.gz && \
-    mv spark-3.1.2-bin-without-hadoop/ /opt/spark && \
-    cd /opt/spark/conf && \
-    mv spark-env.sh.template spark-env.sh
+# Install JDK.
+RUN sudo apt-get install -y openjdk-11-jdk
+ENV JAVA_HOME="/usr/lib/jvm/java-11-openjdk-amd64"
+# Appending our JAVA_HOME so that it would take precedent over bitnami's java home path.
```
Comment (jeremyp-bq) [inline docker/Dockerfile:11]: Minor nit: Prepending.

File: functional-tests/submit-functional-tests.sh [hunk 2]
```diff
@@ -1,8 +1,7 @@
...
-export PATH=$PATH:$SPARK_HOME/bin:$SPARK_HOME/sbin
-start-master.sh -h localhost
-start-worker.sh -h localhost spark://localhost:7077
-spark-submit --master spark://localhost:7077 --driver-memory 2g target/scala-2.12/spark-vertica-connector-functional-tests-assembly-$CONNECTOR_VERSION.jar $1
+
+# Append option -r to the list of args
+args=("-r")
+args+=("$@")
+
+spark-submit --master spark://spark:7077 --driver-memory 2g target/scala-2.12/spark-vertica-connector-functional-tests-assembly-$CONNECTOR_VERSION.jar "${args[@]}"
```
Comment (jeremyp-bq) [inline functional-tests/submit-functional-tests.sh:7]: Would it be worthwhile to make it an option to run locally and/or document how to run locally in the functional-tests README?

File: docker/docker-compose.yml [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
+      args:
+        SPARK: ${SPARK_INSTALL:-latest}
+    image: docker_client_1:${SPARK_CLUSTER:-latest}
     volumes:
       - ./..:/spark-connector
       - ./vertica-hdfs-config/hadoop:/etc/hadoop/conf
     stdin_open: true
     ports:
+#      JVM Remote debug port
       - "5005:5005"
```
Comment (jeremyp-bq) [inline docker/docker-compose.yml:16]: I assume we don't need 5005 here any more, although we might want to add it to the spark master container (assuming it is the master, and not the workers, that exposes the port).

File: docker/docker-compose.yml [hunk 4]
```diff
@@ -1,16 +1,19 @@
 version: "3.9"
 services:
   client:
-    build: .
     container_name: docker_client_1
+    build:
+      context: .
+      args:
+        SPARK: ${SPARK_INSTALL:-latest}
+    image: docker_client_1:${SPARK_CLUSTER:-latest}
```
Comment (jeremyp-bq) [inline docker/docker-compose.yml:9]: What is SPARK_CLUSTER?  Don't see it mentioned anywhere.

File: docker/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+We provide a docker environment for testing and developing. To start, use `./sandbox-clientenv.bat` for Windows 
+or `./sandbox-clientenv.sh` for Linux based system.
+
+The docker environment contains:
+- `docker_client_1`: The client environment. This is where spark application are submitted/executed from
+- `docker_hdfs_1`: The HDFS node acting as the staging file-store
+- `docker_vertica_1`: Our Vertica database
+- `docker_minio_1`: A minio file-store container acting as an S3 file-store 
+- `spark`: A Spark driver for our standalone cluster
+- `spark-worker`: Spark workers
```
Comment (jeremyp-bq) [inline docker/README.md:12]: Nit: spark-worker-* (correct?).

File: docker/docker-compose.yml [hunk 6]
```diff
@@ -22,6 +25,35 @@ services:
       - GCS_SERVICE_EMAIL
       - AWS_ACCESS_KEY_ID
       - AWS_SECRET_ACCESS_KEY
+  spark:
+    image: bitnami/spark:${SPARK_INSTALL:-latest}
+    container_name: spark
+    hostname: spark
+    volumes:
+      - ./..:/spark-connector
```
Comment (jeremyp-bq) [inline docker/docker-compose.yml:33]: Do the master and worker need access to the whole repo?  Or do they just need the HDFS config (or anything at all)?

File: docker/sandbox-clientenv.sh [hunk 7]
```diff
@@ -60,5 +63,8 @@ else
   docker cp ../functional-tests/src/main/resources/3.1.1 docker_hdfs_1:/partitioned
   docker exec docker_hdfs_1 hadoop fs -copyFromLocal /partitioned /3.1.1
   docker exec docker_vertica_1 vsql -c "ALTER DATABASE docker SET MaxClientSessions=100;"
+  docker exec -w /spark-connector/docker/cluster spark-driver /bin/bash -c './start-master.sh'
+  docker exec -w /spark-connector/docker/cluster spark-worker-1 /bin/bash -c './start-worker.sh'
+  docker exec -w /spark-connector/docker/cluster spark-worker-2 /bin/bash -c './start-worker.sh'
```
Comment (jeremyp-bq) [inline docker/sandbox-clientenv.sh:68]: Maybe we should have a loop that executes the command from 1 to WORKERS_COUNT?

File: docker/Dockerfile (hunk 1 above)
Comment (Aryex) [reply docker/Dockerfile:11]: Not minor at all lol. Fixed.

File: functional-tests/submit-functional-tests.sh (hunk 2 above)
Comment (Aryex) [reply functional-tests/submit-functional-tests.sh:7]: My intention is that this script is a shortcut to submit to our cluster. I would assume that people wanting to submit this to their own cluster would probably know who to use `spark-submit` already.

That said, Readme updated to mention submitting outside of our own cluster.

File: docker/docker-compose.yml (hunk 4 above)
Comment (Aryex) [reply docker/docker-compose.yml:9]: Updated. That should be SPARK_INSTALL. It is there to specify a tag of the for our client image.

File: docker/README.md (hunk 5 above)
Comment (Aryex) [reply docker/README.md:12]: updated.

File: docker/sandbox-clientenv.sh (hunk 7 above)
Comment (Aryex) [reply docker/sandbox-clientenv.sh:68]: oops. These should be removed.

File: docker/docker-compose.yml (hunk 3 above)
Comment (Aryex) [reply docker/docker-compose.yml:16]: Actually, we would still want it on the client and not the master node.
Updated the submit script with the option to use debug.

File: functional-tests/submit-functional-tests.sh (hunk 2 above)
Comment (Aryex) [reply functional-tests/submit-functional-tests.sh:7]: The readme edit is removed. See #473

File: docker/docker-compose.yml (hunk 6 above)
Comment (Aryex) [reply docker/docker-compose.yml:33]: Our functional tests are reading files from hardcoded local files, which may make it difficult to run our functional tests on a user's cluster.
Right now, our cluster would mount these folders as a bandaid, but we should address it sometime.
See #473.

Comment (jeremyp-bq) [reply docker/docker-compose.yml:33]: Sounds good, we should make the tests more portable but it will take some effort.

**Review by jeremyp-bq — APPROVED** [review]

**Pull request merged**